### PR TITLE
market_participant_unit_of_work #518

### PR DESCRIPTION
**Pull request by michaelmajgaard**

### CODE REVIEW

File: source/Energinet.DataHub.MarketParticipant.Domain/IUnitOfWorkProvider.cs [hunk 1]
```diff
@@ -0,0 +1,29 @@
...
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+using System.Threading.Tasks;
+
+namespace Energinet.DataHub.MarketParticipant.Domain
+{
+    /// <summary>
+    /// TODO
```
Comment (michaelmajgaard) [inline source/Energinet.DataHub.MarketParticipant.Domain/IUnitOfWorkProvider.cs:20]: Fix summary

File: source/Energinet.DataHub.MarketParticipant.Application/Commands/DispatchEventsCommand.cs [hunk 2]
```diff
@@ -0,0 +1,20 @@
...
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+using MediatR;
+
+namespace Energinet.DataHub.MarketParticipant.Application.Commands
+{
+    public sealed record DispatchEventsCommand : IRequest<DispatchEventsResponse>;
```
Comment (defectiveAi) [inline source/Energinet.DataHub.MarketParticipant.Application/Commands/DispatchEventsCommand.cs:19]: Use Unit instead, then you can murder DispatchEventsResponse-class.

File: source/Energinet.DataHub.MarketParticipant.ApplyDBMigrationsApp/Scripts/U-001/Model/202203061323 Create Domain Event table.sql [hunk 3]
```diff
@@ -0,0 +1,19 @@
+SET ANSI_NULLS ON
```
Comment (defectiveAi) [inline source/Energinet.DataHub.MarketParticipant.ApplyDBMigrationsApp/Scripts/U-001/Model/202203061323 Create Domain Event table.sql:1]: Add the script to all folders/environments.

File: source/Energinet.DataHub.MarketParticipant.Domain/IUnitOfWork.cs [hunk 4]
```diff
@@ -0,0 +1,30 @@
...
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+using System;
+using System.Threading.Tasks;
+
+namespace Energinet.DataHub.MarketParticipant.Domain
+{
+    /// <summary>
+    /// <see cref="IUnitOfWork"/>
```
Comment (defectiveAi) [inline source/Energinet.DataHub.MarketParticipant.Domain/IUnitOfWork.cs:21]: Circular reference.

File: source/Energinet.DataHub.MarketParticipant.Domain/IUnitOfWorkProvider.cs [hunk 5]
```diff
@@ -0,0 +1,29 @@
...
+
+namespace Energinet.DataHub.MarketParticipant.Domain
+{
+    /// <summary>
+    /// Provider for <see cref="IUnitOfWork"/>
+    /// </summary>
+    public interface IUnitOfWorkProvider
+    {
+        /// <summary>
+        /// Creates an initialized a new <see cref="IUnitOfWork"/>
```
Comment (defectiveAi) [inline source/Energinet.DataHub.MarketParticipant.Domain/IUnitOfWorkProvider.cs:25]: It's a me, a Mario!

File: source/Energinet.DataHub.MarketParticipant.Domain/Model/DomainEvent.cs [hunk 6]
```diff
@@ -0,0 +1,52 @@
...
+            IsSent = isSent;
+        }
+
+        public DomainEventId Id { get; } = new DomainEventId(0);
+        public Guid DomainObjectId { get; }
+        public string DomainObjectType { get; }
+        public bool IsSent { get; private set; }
+        public IIntegrationEvent IntegrationEvent { get; }
+
+        public static DomainEvent Restore(DomainEventId id, Guid domainObjectId, string domainObjectType, bool isSent, IIntegrationEvent integrationEvent)
```
Comment (defectiveAi) [inline source/Energinet.DataHub.MarketParticipant.Domain/Model/DomainEvent.cs:42]: Just make the ctor public. This method does nothing for readability, is just more code. It is obvious from the two ctor overloads, which one is for new'ing an object and which one is for restoring it.

File: source/Energinet.DataHub.MarketParticipant.EntryPoint.Organization/Energinet.DataHub.MarketParticipant.EntryPoint.Organization.csproj [hunk 7]
```diff
@@ -31,6 +31,7 @@ limitations under the License.
     <PackageReference Include="MediatR" Version="10.0.1" />
     <PackageReference Include="Microsoft.Azure.Functions.Worker" Version="1.6.0" />
     <PackageReference Include="Microsoft.Azure.Functions.Worker.Extensions.Http" Version="3.0.13" />
+    <PackageReference Include="Microsoft.Azure.Functions.Worker.Extensions.Timer" Version="4.1.0" />
```
Comment (defectiveAi) [inline source/Energinet.DataHub.MarketParticipant.EntryPoint.Organization/Energinet.DataHub.MarketParticipant.EntryPoint.Organization.csproj:34]: This should eventually be moved into a new EntryPoint. But not necessarily in this PR.

**Review by mbusk88 — CHANGES_REQUESTED** [review]
See comment

File: source/Energinet.DataHub.MarketParticipant.Infrastructure/Persistence/Repositories/DomainEventRepository.cs [hunk 8]
```diff
@@ -0,0 +1,114 @@
...
+    public sealed class DomainEventRepository : IDomainEventRepository
+    {
+        private readonly IMarketParticipantDbContext _context;
+        private readonly IDictionary<string, Type> _domainEventTypesByName;
+        private readonly JsonSerializerOptions _jsonSerializerOptions = new JsonSerializerOptions { PropertyNamingPolicy = JsonNamingPolicy.CamelCase };
+
+        public DomainEventRepository(IMarketParticipantDbContext context)
+        {
+            _context = context;
+            _domainEventTypesByName = typeof(IIntegrationEvent).Assembly.GetTypes().Where(x => x.IsAssignableTo(typeof(IIntegrationEvent))).ToDictionary(x => x.Name);
```
Comment (defectiveAi) [inline source/Energinet.DataHub.MarketParticipant.Infrastructure/Persistence/Repositories/DomainEventRepository.cs:41]: Didn't you register the collection? Just ask for IEnumerable<IIntegrationEvent>?

File: source/Energinet.DataHub.MarketParticipant.Infrastructure/Persistence/Repositories/DomainEventRepository.cs [hunk 9]
```diff
@@ -0,0 +1,114 @@
...
+
+        public async Task UpdateAsync(DomainEvent domainEvent)
+        {
+            Guard.ThrowIfNull(domainEvent, nameof(domainEvent));
+
+            var q = from x in _context.DomainEvents.AsQueryable()
+                    where x.Id == domainEvent.Id.Value
+                    select x;
+
+            var entity = await q.FirstOrDefaultAsync().ConfigureAwait(false);
```
Comment (defectiveAi) [inline source/Energinet.DataHub.MarketParticipant.Infrastructure/Persistence/Repositories/DomainEventRepository.cs:75]: You can use Find(..), which is more efficient.

File: source/Energinet.DataHub.MarketParticipant.IntegrationTests/Repositories/DomainEventRepositoryTests.cs [hunk 10]
```diff
@@ -0,0 +1,116 @@
...
+            await using var scope = host.BeginScope();
+            await using var context = _fixture.DatabaseManager.CreateDbContext();
+            var repository = new DomainEventRepository(context);
+
+            // act
+            await repository
+                .InsertAsync(new DomainEvent(Guid.NewGuid(), nameof(Organization), new OrganizationChangedIntegrationEvent { OrganizationId = Guid.NewGuid(), ActorId = Guid.NewGuid(), Gln = "gln", Name = "name" }))
+                .ConfigureAwait(false);
+
+            // assert
```
Comment (defectiveAi) [inline source/Energinet.DataHub.MarketParticipant.IntegrationTests/Repositories/DomainEventRepositoryTests.cs:53]: Get event to see if it is there?

File: source/Energinet.DataHub.MarketParticipant.IntegrationTests/Repositories/DomainEventRepositoryTests.cs [hunk 11]
```diff
@@ -0,0 +1,116 @@
...
+        }
```
[the remark this answers is not in the document]
Comment (michaelmajgaard) [reply source/Energinet.DataHub.MarketParticipant.IntegrationTests/Repositories/DomainEventRepositoryTests.cs:55]: fixed

**Review by defectiveAi — CHANGES_REQUESTED** [review]

File: source/Energinet.DataHub.MarketParticipant.Infrastructure/Persistence/Repositories/DomainEventRepository.cs (hunk 8 above)
Comment (michaelmajgaard) [reply source/Energinet.DataHub.MarketParticipant.Infrastructure/Persistence/Repositories/DomainEventRepository.cs:41]: These are not what you think they are. Look again.

File: source/Energinet.DataHub.MarketParticipant.ApplyDBMigrationsApp/Scripts/U-001/Model/202203061323 Create Domain Event table.sql (hunk 3 above)
Comment (michaelmajgaard) [reply source/Energinet.DataHub.MarketParticipant.ApplyDBMigrationsApp/Scripts/U-001/Model/202203061323 Create Domain Event table.sql:1]: I fear no man, but that thing.. It scares me.

File: source/Energinet.DataHub.MarketParticipant.Domain/IUnitOfWork.cs (hunk 4 above)
Comment (michaelmajgaard) [reply source/Energinet.DataHub.MarketParticipant.Domain/IUnitOfWork.cs:21]: I'm funny

File: source/Energinet.DataHub.MarketParticipant.Domain/IUnitOfWorkProvider.cs (hunk 5 above)
Comment (michaelmajgaard) [reply source/Energinet.DataHub.MarketParticipant.Domain/IUnitOfWorkProvider.cs:25]: You're not funny

File: source/Energinet.DataHub.MarketParticipant.Infrastructure/Persistence/Repositories/DomainEventRepository.cs (hunk 8 above)
Comment (defectiveAi) [reply source/Energinet.DataHub.MarketParticipant.Infrastructure/Persistence/Repositories/DomainEventRepository.cs:41]: Ah, is not dispatchers. My bad.

File: source/Energinet.DataHub.MarketParticipant.Domain/IUnitOfWork.cs (hunk 4 above)
Comment (defectiveAi) [reply source/Energinet.DataHub.MarketParticipant.Domain/IUnitOfWork.cs:21]: Nuhuh... Spent 6 minutes following the cref, until brain crashed.

File: source/Energinet.DataHub.MarketParticipant.Application/Commands/DispatchEventsCommand.cs (hunk 2 above)
Comment (michaelmajgaard) [reply source/Energinet.DataHub.MarketParticipant.Application/Commands/DispatchEventsCommand.cs]: fixed

File: source/Energinet.DataHub.MarketParticipant.ApplyDBMigrationsApp/Scripts/U-001/Model/202203061323 Create Domain Event table.sql (hunk 3 above)
Comment (michaelmajgaard) [reply source/Energinet.DataHub.MarketParticipant.ApplyDBMigrationsApp/Scripts/U-001/Model/202203061323 Create Domain Event table.sql:1]: fixed

File: source/Energinet.DataHub.MarketParticipant.Domain/IUnitOfWork.cs (hunk 4 above)
Comment (michaelmajgaard) [reply source/Energinet.DataHub.MarketParticipant.Domain/IUnitOfWork.cs]: fixed

File: source/Energinet.DataHub.MarketParticipant.Domain/IUnitOfWorkProvider.cs (hunk 5 above)
Comment (michaelmajgaard) [reply source/Energinet.DataHub.MarketParticipant.Domain/IUnitOfWorkProvider.cs]: fixed

File: source/Energinet.DataHub.MarketParticipant.Domain/Model/DomainEvent.cs (hunk 6 above)
Comment (michaelmajgaard) [reply source/Energinet.DataHub.MarketParticipant.Domain/Model/DomainEvent.cs]: fixed

File: source/Energinet.DataHub.MarketParticipant.Infrastructure/Persistence/Repositories/DomainEventRepository.cs (hunk 9 above)
Comment (michaelmajgaard) [reply source/Energinet.DataHub.MarketParticipant.Infrastructure/Persistence/Repositories/DomainEventRepository.cs]: fixed

File: source/Energinet.DataHub.MarketParticipant.IntegrationTests/Repositories/DomainEventRepositoryTests.cs (hunk 10 above)
Comment (michaelmajgaard) [reply source/Energinet.DataHub.MarketParticipant.IntegrationTests/Repositories/DomainEventRepositoryTests.cs:53]: fixed

File: source/Energinet.DataHub.MarketParticipant.Infrastructure/Persistence/Repositories/DomainEventRepository.cs [hunk 12]
```diff
@@ -0,0 +1,114 @@
...
+            await _context.SaveChangesAsync().ConfigureAwait(false);
+
+            return new DomainEventId(entity.Id);
+        }
+
+        public async Task UpdateAsync(DomainEvent domainEvent)
+        {
+            Guard.ThrowIfNull(domainEvent, nameof(domainEvent));
+
+            var q = from x in _context.DomainEvents.AsQueryable()
```
Comment (defectiveAi) [inline source/Energinet.DataHub.MarketParticipant.Infrastructure/Persistence/Repositories/DomainEventRepository.cs:71]: q not used anymore?

Comment (michaelmajgaard) [reply source/Energinet.DataHub.MarketParticipant.Infrastructure/Persistence/Repositories/DomainEventRepository.cs]: woops, fixed

**Review by mbusk88 — APPROVED** [review]

**Review by defectiveAi — DISMISSED** [review]

**Review by defectiveAi — APPROVED** [review]

**Review by FirestarJes — APPROVED** [review]

**Pull request merged**